added timestamp proposal check so that it does not go beyond timewindow

pfi79 opened the pull request:
Added an auth filter that checks the timestamp of the proposal so that it doesn't differ too much from the peer time using the peer.authentication.timewindow parameter.

Also made it so that gateway calling Proposal locally does not pass auth filters

pfi79 committed:
added timestamp proposal check so that it does not go beyond timewindow

Signed-off-by: Fedor Partanskiy <fredprtnsk@gmail.com>

C0rWin commented:
thanks @pfi79, could you please add integration tests also?

pfi79 commented:
> thanks @pfi79, could you please add integration tests also?

Okay, I will.

yacovm commented:
Do clients (SDKs, gateway pseudo-SDK) actually set the timestamp in the channel header? Let's be sure we're not breaking all clients with this PR :-)

yacovm commented on sampleconfig/core.yaml:
```diff
@@ -454,13 +454,11 @@ peer:
     handlers:
         authFilters:
-          -
-            name: DefaultAuth
```
Looking at this, I'm not even sure we should have activated the expiration check by default.
So, I am not sure we should have this new check activated by default.

Perhaps let it be commented out?

pfi79 replied:
1. The DefaultAuth and ExpirationCheck filters were enabled by default. I don't know how long it's been on.

2. I think that enabling filters by default makes the fabric safer for newcomers who are just starting to use and familiarize themselves with it.

3. Confident professionals can comment these filters consciously, realizing what it can lead to.

yacovm replied:
In that case I'd like to hear other people's thoughts on this. 

@C0rWin , or @denyeart  wdyt ?

denyeart replied:
I think this is a great feature for v3, thank you @pfi79.
Since it is a major release, I think it is completely acceptable to add a new check, as long as we remember to mention it in the release notes.

C0rWin replied:
I think I do agree with @denyeart. This is a good feature as it will eliminate the issues with newcomers, and given v3.0 already brings some breaking changes, IMO,  this should not be a problem enabling it.

yacovm commented on core/handlers/library/registry.go:
```diff
@@ -73,6 +75,8 @@ func InitRegistry(c Config) Registry {
 
 // loadHandlers loads the configured handlers
 func (r *registry) loadHandlers(c Config) {
+	r.authenticationTimeWindow = c.AuthenticationTimeWindow
```
I think that `InitRegistry()` is the right place to initialize it. Not here.

pfi79 replied:
fix

pfi79 commented:
> Do clients (SDKs, gateway pseudo-SDK) actually set the timestamp in the channel header? Let's be sure we're not breaking all clients with this PR :-)

This is the time that we then get inside the chaincode using stub.GetTxTimestamp()

yacovm commented:
> > Do clients (SDKs, gateway pseudo-SDK) actually set the timestamp in the channel header? Let's be sure we're not breaking all clients with this PR :-)
> 
> This is the time that we then get inside the chaincode using stub.GetTxTimestamp()

Yeah I'm asking whether all client implementations correctly set it. We should check this, otherwise we're breaking the clients.

pfi79 commented:
> > > Do clients (SDKs, gateway pseudo-SDK) actually set the timestamp in the channel header? Let's be sure we're not breaking all clients with this PR :-)
> > 
> > 
> > This is the time that we then get inside the chaincode using stub.GetTxTimestamp()
> 
> Yeah I'm asking whether all client implementations correctly set it. We should check this, otherwise we're breaking the clients.

Then I have a counter question.
What do you call "properly installed" and "improperly installed"?

yacovm commented:
> > > > Do clients (SDKs, gateway pseudo-SDK) actually set the timestamp in the channel header? Let's be sure we're not breaking all clients with this PR :-)
> > > 
> > > 
> > > This is the time that we then get inside the chaincode using stub.GetTxTimestamp()
> > 
> > 
> > Yeah I'm asking whether all client implementations correctly set it. We should check this, otherwise we're breaking the clients.
> 
> Then I have a counter question. What do you call "properly installed" and "improperly installed"?

well for instance, a client may not initialize the timestamp field at all.

pfi79 commented:
> well for instance, a client may not initialize the timestamp field at all.

That's great, I've never seen one of those.

Then they don't use stub.GetTxTimestamp() in their chaincode. No way, especially for selecting the code execution path (if....else).  
They can comment out this filter in core.yaml.

yacovm commented:
> > well for instance, a client may not initialize the timestamp field at all.
> 
> That's great, I've never seen one of those.
> 
> Then they don't use stub.GetTxTimestamp() in their chaincode. No way, especially for selecting the code execution path (if....else). They can comment out this filter in core.yaml.

No I mean the code that the client SDK executes, the code that is found in the official Fabric client SDK, not the code that the user implements as part of the solution. It doesn't matter if they use the chaincode stub API, as the proposal won't reach that far with the new filter. And the user and the deployer of the peer may be different people, so we should verify the client SDK code in multiple implementations sets this field.

pfi79 commented:
> > > well for instance, a client may not initialize the timestamp field at all.
> > 
> > 
> > That's great, I've never seen one of those.
> > Then they don't use stub.GetTxTimestamp() in their chaincode. No way, especially for selecting the code execution path (if....else). They can comment out this filter in core.yaml.
> 
> No I mean the code that the client SDK executes, the code that is found in the official Fabric client SDK, not the code that the user implements as part of the solution. It doesn't matter if they use the chaincode stub API, as the proposal won't reach that far with the new filter. And the user and the deployer of the peer may be different people, so we should verify the client SDK code in multiple implementations sets this field.

I'll check

pfi79 commented:
> > > well for instance, a client may not initialize the timestamp field at all.
> > 
> > 
> > That's great, I've never seen one of those.
> > Then they don't use stub.GetTxTimestamp() in their chaincode. No way, especially for selecting the code execution path (if....else). They can comment out this filter in core.yaml.
> 
> No I mean the code that the client SDK executes, the code that is found in the official Fabric client SDK, not the code that the user implements as part of the solution. It doesn't matter if they use the chaincode stub API, as the proposal won't reach that far with the new filter. And the user and the deployer of the peer may be different people, so we should verify the client SDK code in multiple implementations sets this field.

checked:
- fabric-gateway
- fabric-sdk-java
- fabric-sdk-node
- fabric-sdk-go
- fabric-admin-sdk

everyone fills in the timestamp field

Forgetting someone?

yacovm commented:
> > > > well for instance, a client may not initialize the timestamp field at all.
> > > 
> > > 
> > > That's great, I've never seen one of those.
> > > Then they don't use stub.GetTxTimestamp() in their chaincode. No way, especially for selecting the code execution path (if....else). They can comment out this filter in core.yaml.
> > 
> > 
> > No I mean the code that the client SDK executes, the code that is found in the official Fabric client SDK, not the code that the user implements as part of the solution. It doesn't matter if they use the chaincode stub API, as the proposal won't reach that far with the new filter. And the user and the deployer of the peer may be different people, so we should verify the client SDK code in multiple implementations sets this field.
> 
> checked:
> 
>     * fabric-gateway
> 
>     * fabric-sdk-java
> 
>     * fabric-sdk-node
> 
>     * fabric-sdk-go
> 
>     * fabric-admin-sdk
> 
> 
> everyone fills in the timestamp field
> 
> Forgetting someone?

Can you please link to the corresponding lines, if that's not too much trouble?

pfi79 commented:
> > > > > well for instance, a client may not initialize the timestamp field at all.
> > > > 
> > > > 
> > > > That's great, I've never seen one of those.
> > > > Then they don't use stub.GetTxTimestamp() in their chaincode. No way, especially for selecting the code execution path (if....else). They can comment out this filter in core.yaml.
> > > 
> > > 
> > > No I mean the code that the client SDK executes, the code that is found in the official Fabric client SDK, not the code that the user implements as part of the solution. It doesn't matter if they use the chaincode stub API, as the proposal won't reach that far with the new filter. And the user and the deployer of the peer may be different people, so we should verify the client SDK code in multiple implementations sets this field.
> > 
> > 
> > checked:
> > ```
> > * fabric-gateway
> > 
> > * fabric-sdk-java
> > 
> > * fabric-sdk-node
> > 
> > * fabric-sdk-go
> > 
> > * fabric-admin-sdk
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > everyone fills in the timestamp field
> > Forgetting someone?
> 
> Can you please link to the corresponding lines, if that's not too much trouble?

we are interested in ChannelHeader only during the formation of Proposal
- fabric-gateway
[go](https://github.com/hyperledger/fabric-gateway/blob/a4aef4da2aa565ba296a76a0fff55083c4db8ebd/pkg/client/proposalbuilder.go#L118) 
[java](https://github.com/hyperledger/fabric-gateway/blob/a4aef4da2aa565ba296a76a0fff55083c4db8ebd/java/src/main/java/org/hyperledger/fabric/client/ProposalBuilder.java#L130)
[node](https://github.com/hyperledger/fabric-gateway/blob/a4aef4da2aa565ba296a76a0fff55083c4db8ebd/node/src/proposalbuilder.ts#L96)
- fabric-sdk-java
[this](https://github.com/hyperledger/fabric-sdk-java/blob/5a5259d77cf27a91f99407a5d3807ca6e664d618/src/main/java/org/hyperledger/fabric/sdk/transaction/ProposalBuilder.java#L151) [this](https://github.com/hyperledger/fabric-sdk-java/blob/5a5259d77cf27a91f99407a5d3807ca6e664d618/src/main/java/org/hyperledger/fabric/sdk/transaction/ProposalBuilder.java#L129) [this](https://github.com/hyperledger/fabric-sdk-java/blob/5a5259d77cf27a91f99407a5d3807ca6e664d618/src/main/java/org/hyperledger/fabric/sdk/transaction/ProposalBuilder.java#L130) and [this](https://github.com/hyperledger/fabric-sdk-java/blob/main/src/main/java/org/hyperledger/fabric/sdk/transaction/ProtoUtils.java#L94)
- fabric-sdk-node
[this](https://github.com/hyperledger/fabric-sdk-node/blob/fa3df2b69a7323fffe127d9560491257f8e097fe/fabric-common/lib/Proposal.js#L292) [this](https://github.com/hyperledger/fabric-sdk-node/blob/fa3df2b69a7323fffe127d9560491257f8e097fe/fabric-common/lib/Channel.js#L506)
- fabric-sdk-go
[this](https://github.com/hyperledger/fabric-sdk-go/blob/5d6ca326e01b057f4f1b3233e70a36c7623fae6f/internal/github.com/hyperledger/fabric/protoutil/proputils.go#L98)
- fabric-admin-sdk
[node1](https://github.com/hyperledger/fabric-admin-sdk/blob/f896a39f3e3e20ee003cb9fbd712a742bd97c3a8/ts/formatter/lib/proto/proposal-builder.ts#L66) [node2](https://github.com/hyperledger/fabric-admin-sdk/blob/f896a39f3e3e20ee003cb9fbd712a742bd97c3a8/ts/formatter/lib/proto/channel-builder.ts#L32)
[go](https://github.com/hyperledger/fabric-admin-sdk/blob/f896a39f3e3e20ee003cb9fbd712a742bd97c3a8/pkg/internal/proposal/builder.go#L121)

denyeart commented on sampleconfig/core.yaml:
```diff
@@ -454,13 +454,11 @@ peer:
-            name: ExpirationCheck    # This filter checks identity x509 certificate expiration
+          - name: DefaultAuth
+          - name: ExpirationCheck # This filter checks identity x509 certificate expiration
+          - name: TimeWindowCheck # This filter checks Timestamp in TimeWindow
```
Please add a more descriptive comment, mentioning that the client provided timestamp will be checked against the peer's timestamp, using the timewindow configured above in `peer.authentication.timewindow`.

Also, in the `peer.authentication.timewindow`config comment above, indicate that the timewindow value is used for delivery service and endorsement service (if the authFilter is enabled).  I can update the comment in release-2.5 branch to indicate that the timewindow was only used on delivery path in that release.

pfi79 replied:
added

denyeart commented on core/handlers/auth/filter/timewindow.go:
```diff
@@ -0,0 +1,68 @@
+	now := time.Now().UTC()
+
+	if timeProposal.Add(timeWindow).Before(now) || timeProposal.Add(-timeWindow).After(now) {
+		return errors.New("request unauthorized due to incorrect timestamp")
```
Please add a more descriptive error message, including the client timestamp, peer timestamp, and configured timewindow. Mention peer.authorization.timewindow property by name so that users can tie it back to the config in core.yaml.

pfi79 replied:
added

pfi79 commented:
> thanks @pfi79, could you please add integration tests also?

added an integration test

also found that Proposal was bypassing auth filters locally via gateway
fixed

denyeart commented:
@pfi79 All pull requests should have some description text, even if it is short and sweet.

pfi79 commented:
> @pfi79 All pull requests should have some description text, even if it is short and sweet.

added

pfi79 commented:
@C0rWin @yacovm @denyeart 
I think I've corrected all the comments. 
Please take a look and if I forgot something, please indicate it.

By the way, since yesterday the pipelines stopped working. 
Now there are 4 frozen pipelines

denyeart reviewed (approved):
Looks good to me, Iet's see if there are any other final comments...

C0rWin commented:
> Looks good to me, Iet's see if there are any other final comments...

I think we need to take to properly document it to make sure there are nor further confusions. I am good with the change.
